Deployment/Jenkins/agent/agent-instance.groovy

```diff
@@ -0,0 +1,258 @@
+                                    "ParameterValue": "${env.PyStache}"
+                                },
+                                {
+                                    "ParameterKey": "SecurityGroupIds",
+                                    "ParameterValue": "${env.SecurityGroupIds}"
+                                },
+                                {
+                                    "ParameterKey": "SubnetId",
+                                    "ParameterValue": "${env.SubnetId}"
+                                },
+                                {
+                                    "ParameterKey": "ZapAgentUrl",
+                                    "ParameterValue": "${env.ZapAgentUrl}"
+                                }
+                            ]
+                        /
+                }
+            }
+        }
+        stage ('Prepare AWS Environment') {
+            options {
+                timeout(time: 1, unit: 'HOURS')
+            }
+            steps {
+                withCredentials(
+                    [
+                        [$class: 'AmazonWebServicesCredentialsBinding', accessKeyVariable: 'AWS_ACCESS_KEY_ID', credentialsId: "${AwsCred}", secretKeyVariable: 'AWS_SECRET_ACCESS_KEY'],
+                        sshUserPrivateKey(credentialsId: "${GitCred}", keyFileVariable: 'SSH_KEY_FILE', passphraseVariable: 'SSH_KEY_PASS', usernameVariable: 'SSH_KEY_USER')
+                    ]
+                ) {
+                    sh '''#!/bin/bash
+                        echo "Attempting to delete any active ${CfnStackRoot} stacks... "
+                        aws --region "${AwsRegion}" cloudformation delete-stack --stack-name "${CfnStackRoot}"
+
+                        aws cloudformation wait stack-delete-complete --stack-name ${CfnStackRoot} --region ${AwsRegion}
+                    '''
+                }
+            }
+        }
+        stage ('Launch Jenkins Master Stack') {
+            options {
+                timeout(time: 1, unit: 'HOURS')
+            }
+            steps {
+                withCredentials(
+                    [
+                        [$class: 'AmazonWebServicesCredentialsBinding', accessKeyVariable: 'AWS_ACCESS_KEY_ID', credentialsId: "${AwsCred}", secretKeyVariable: 'AWS_SECRET_ACCESS_KEY'],
+                        sshUserPrivateKey(credentialsId: "${GitCred}", keyFileVariable: 'SSH_KEY_FILE', passphraseVariable: 'SSH_KEY_PASS', usernameVariable: 'SSH_KEY_USER')
+                    ]
+                ) {
+                    sh '''#!/bin/bash -xe
+                        printf 'export SIGNED_URL="%s"' "\$( aws s3 presign ${ChainScriptUrl} )" > /tmp/SIGNED_URL.txt
+                        source /tmp/SIGNED_URL.txt
+                        printenv | sort
+                        echo "Attempting to create stack ${CfnStackRoot}..."
+                        aws --region "${AwsRegion}" cloudformation create-stack --stack-name "${CfnStackRoot}" \
```

Comment:
You'll need to fix the `--stack-name`. Haven't looked at the other pipeline file, yet, but if both of them are setting their `--stack-name` the same way and two people select the same value for this parameter, "bad things will happen". I would probably change this from:

~~~
--stack-name "${CfnStackRoot}"
~~~

to

~~~
--stack-name "${CfnStackRoot}-AGENT"
~~~

Also, be aware that even with the above change, if you wanted to deploy more than one, concurrent agent, you'd need to ensure that either `${CfnStackRoot}` is unique per agent-node or you would need to add further uniqueness to this string.

Whatever you change this to, make sure that your deletion-stage matches.

Reply:
Resolved the stage-name suggestion but not the incipient stack-name collision problem?

Reply:
Added ${BUILD_NUMBER} here as well as to the Master